File: app/src/main/java/com/onemoresecret/KeyImportFragment.java
package com.onemoresecret;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.view.MenuProvider;
import androidx.fragment.app.Fragment;
import androidx.navigation.fragment.NavHostFragment;

import com.onemoresecret.crypto.AESUtil;
import com.onemoresecret.crypto.AesKeyAlgorithm;
import com.onemoresecret.crypto.AesTransformation;
import com.onemoresecret.crypto.CryptographyManager;
import com.onemoresecret.crypto.MessageComposer;
import com.onemoresecret.databinding.FragmentKeyImportBinding;

import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.security.KeyStoreException;
import java.security.interfaces.RSAPublicKey;
import java.util.Arrays;

import javax.crypto.Cipher;
import javax.crypto.spec.IvParameterSpec;

/**
 * Key import Fragment.
 */
public class KeyImportFragment extends Fragment {
    private FragmentKeyImportBinding binding;
    private static final String TAG = KeyImportFragment.class.getSimpleName();
    private final CryptographyManager cryptographyManager = new CryptographyManager();

    private final KeyImportMenu menuProvider = new KeyImportMenu();

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        requireActivity().addMenuProvider(menuProvider);

        assert getArguments() != null;

        try (OmsDataInputStream dataInputStream = new OmsDataInputStream(new ByteArrayInputStream(getArguments().getByteArray("MESSAGE")))) {
            // (1) Application ID
            var applicationId = dataInputStream.readUnsignedShort();
            if (applicationId != MessageComposer.APPLICATION_AES_ENCRYPTED_PRIVATE_KEY_TRANSFER)
                throw new IllegalArgumentException("wrong applicationId: " + applicationId);

            // (2) alias
            var alias = dataInputStream.readString();
            Log.d(TAG, "alias: " + alias);
            binding.editTextKeyAlias.setText(alias);
            // --- AES parameter ---

            // (3) salt
            var salt = dataInputStream.readByteArray();
            Log.d(TAG, "salt: " + Util.byteArrayToHex(salt));

            // (4) IV
            var iv = dataInputStream.readByteArray();
            Log.d(TAG, "IV: " + Util.byteArrayToHex(iv));

            // (5) AES transformation index
            var aesTransformation = AesTransformation.values()[dataInputStream.readUnsignedShort()].transformation;
            Log.d(TAG, "cipher algorithm: " + aesTransformation);

            // (6) key algorithm index
            var aesKeyAlg = AesKeyAlgorithm.values()[dataInputStream.readUnsignedShort()].keyAlgorithm;
            Log.d(TAG, "AES key algorithm: " + aesKeyAlg);

            // (7) key length
            var aesKeyLength = dataInputStream.readUnsignedShort();
            Log.d(TAG, "AES key length: " + aesKeyLength);

            // (8) AES iterations
            var iterations = dataInputStream.readUnsignedShort();
            Log.d(TAG, "iterations: " + iterations);

            // --- Encrypted part ---

            // (9) cipher text
            var cipherText = dataInputStream.readByteArray();
            Log.d(TAG, cipherText.length + " bytes cipher text read");

            binding.editTextKeyAlias.setText(alias);

            binding.btnDecrypt.setOnClickListener(e ->
                    new Thread(() ->
                            onPasswordEntry(salt,
                                    iv,
                                    cipherText,
                                    aesTransformation,
                                    aesKeyAlg,
                                    aesKeyLength,
                                    iterations)).start()
            );
        } catch (IOException ex) {
            ex.printStackTrace();
            Toast.makeText(getContext(), getString(R.string.wrong_message_format), Toast.LENGTH_LONG).show();
            NavHostFragment.findNavController(this).popBackStack();
        }
    }

    private void onPasswordEntry(
            byte[] salt,
            byte[] iv,
            byte[] cipherText,
            String aesTransformation,
            String keyAlg,
            int keyLength,
            int iterations) {

        try {
            //try decrypt
            var secretKey = AESUtil.getKeyFromPassword(
                    binding.editTextPassphrase.getText().toString().toCharArray(),
                    salt,
                    keyAlg,
                    keyLength,
                    iterations);

            var bArr = AESUtil.process(
                    Cipher.DECRYPT_MODE,
                    cipherText,
                    secretKey,
                    new IvParameterSpec(iv),
                    aesTransformation);

            try (OmsDataInputStream dataInputStream = new OmsDataInputStream(new ByteArrayInputStream(bArr))) {
                // (9.1) - private key material
                var privateKeyMaterial = dataInputStream.readByteArray();

                // (9.2) - public key material
                var publicKeyMaterial = dataInputStream.readByteArray();

                var keyPair = CryptographyManager.restoreKeyPair(privateKeyMaterial, publicKeyMaterial);

                var publicKey = (RSAPublicKey) keyPair.getPublic();

                var fingerprintBytes = CryptographyManager.getFingerprint(publicKey);
                var fingerprint = Util.byteArrayToHex(fingerprintBytes);

                requireContext().getMainExecutor().execute(() -> {
                    binding.textFingerprintData.setText(fingerprint);

                    //check alias
                    binding.editTextKeyAlias.addTextChangedListener(getTextWatcher(fingerprintBytes));

                    validateAlias(fingerprintBytes);

                    binding.btnSave.setEnabled(true);

                    binding.btnSave.setOnClickListener(e ->
                            new Thread(() -> {
                                try {
                                    //delete other keys with the same fingerprint
                                    var sameFingerprint = cryptographyManager.getByFingerprint(fingerprintBytes);
                                    sameFingerprint.forEach(a -> {
                                        try {
                                            cryptographyManager.deleteKey(a);
                                        } catch (KeyStoreException ex) {
                                            ex.printStackTrace();
                                        }
                                    });

                                    var keyAlias = binding.
                                            editTextKeyAlias.
                                            getText().
                                            toString().
                                            trim();
                                    cryptographyManager.importKey(
                                            keyAlias,
                                            keyPair,
                                            requireContext());
                                    requireContext().getMainExecutor().execute(
                                            () -> {
                                                Toast.makeText(this.getContext(),
                                                        "Private key '" + keyAlias + "' successfully imported",
                                                        Toast.LENGTH_LONG).show();
                                                NavHostFragment.findNavController(KeyImportFragment.this).popBackStack();
                                            });

                                } catch (Exception ex) {
                                    ex.printStackTrace();
                                    requireContext().getMainExecutor().execute(
                                            () -> Toast.makeText(this.getContext(),
                                                    ex.getMessage(),
                                                    Toast.LENGTH_LONG).show());
                                }
                            }).start()
                    );
                });
            } catch (IOException ex) {
                ex.printStackTrace();
            }
        } catch (Exception ex) {
            ex.printStackTrace();
            requireContext().getMainExecutor().execute(
                    () -> Toast.makeText(this.getContext(),
                            "Could not decrypt. Wrong passphrase?",
                            Toast.LENGTH_LONG).show());
        }
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentKeyImportBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        requireActivity().removeMenuProvider(menuProvider);
        binding = null;
    }

    private void validateAlias(byte[] fingerprintNew) {
        try {
            var alias = binding.editTextKeyAlias.getText().toString();
            Log.d(TAG, "alias: " + alias);
            String warning = null;

            if (cryptographyManager.keyStore.containsAlias(alias)) {
                var publicKey = (RSAPublicKey) cryptographyManager.getCertificate(alias).getPublicKey();
                var fingerprint = CryptographyManager.getFingerprint(publicKey);
                if (!Arrays.equals(fingerprint, fingerprintNew)) {
                    warning = String.format(getString(R.string.warning_alias_exists), Util.byteArrayToHex(fingerprint));
                }
            }

            var sameFingerprint = cryptographyManager.getByFingerprint(fingerprintNew);
            if (!sameFingerprint.isEmpty()) {
                warning = String.format(getString(R.string.warning_same_fingerprint), sameFingerprint.get(0));
            }

            var _warning = warning;

            requireContext().getMainExecutor().execute(() -> {
                binding.txtWarnings.setText(_warning == null ? "" : _warning);
                binding.txtWarnings.setVisibility(_warning == null ? View.GONE : View.VISIBLE);
            });

        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private TextWatcher getTextWatcher(byte[] fingerprintBytes) {
        return new TextWatcher() {

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {
            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
            }

            @Override
            public void afterTextChanged(Editable s) {
                validateAlias(fingerprintBytes);
            }
        };
    }

    private class KeyImportMenu implements MenuProvider {

        @Override
        public void onCreateMenu(@NonNull Menu menu, @NonNull MenuInflater menuInflater) {
            menuInflater.inflate(R.menu.menu_help, menu);
        }

        @Override
        public boolean onMenuItemSelected(@NonNull MenuItem menuItem) {
            if (menuItem.getItemId() == R.id.menuItemHelp) {
                Util.openUrl(R.string.key_import_md_url, requireContext());
            } else {
                return false;
            }
            return true;
        }
    }

}

File: app/src/main/java/com/onemoresecret/KeyImportFragment.kt
package com.onemoresecret

import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.core.view.MenuProvider
import androidx.fragment.app.Fragment
import com.onemoresecret.Util.byteArrayToHex
import com.onemoresecret.Util.discardBackStack
import com.onemoresecret.Util.openUrl
import com.onemoresecret.Util.printStackTrace
import com.onemoresecret.composable.KeyImportScreen
import com.onemoresecret.composable.OneMoreSecretTheme
import com.onemoresecret.crypto.AESUtil.getAesKeyMaterialFromPassword
import com.onemoresecret.crypto.AESUtil.process
import com.onemoresecret.crypto.AesKeyAlgorithm
import com.onemoresecret.crypto.AesTransformation
import com.onemoresecret.crypto.CryptographyManager
import com.onemoresecret.crypto.CryptographyManager.Companion.PROP_KEYSTORE
import com.onemoresecret.crypto.MessageComposer
import com.onemoresecret.crypto.RSAUtil.getFingerprint
import com.onemoresecret.crypto.RSAUtil.restorePublicKey
import java.io.ByteArrayInputStream
import java.io.IOException
import java.security.interfaces.RSAPublicKey
import java.util.Arrays
import java.util.Objects
import javax.crypto.Cipher

/**
 * Key import Fragment.
 */
class KeyImportFragment : Fragment() {
    private val cryptographyManager = CryptographyManager()
    private val menuProvider = KeyImportMenu()
    private lateinit var preferences: SharedPreferences

    private var keyAlias by mutableStateOf("")
    private var passphrase by mutableStateOf("")
    private var fingerprint by mutableStateOf("")
    private var warning by mutableStateOf("")
    private var saveEnabled by mutableStateOf(false)

    private var decryptedPrivateKeyMaterial: ByteArray? = null
    private var decryptedPublicKeyMaterial: ByteArray? = null
    private var decryptedFingerprintBytes: ByteArray? = null
    private var encryptedPayload: EncryptedPayload? = null

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        preferences = requireActivity().getPreferences(Context.MODE_PRIVATE)
        requireActivity().addMenuProvider(menuProvider)

        try {
            OmsDataInputStream(ByteArrayInputStream(requireArguments().getByteArray("MESSAGE"))).use { dataInputStream ->
                // (1) Application ID
                val applicationId = dataInputStream.readUnsignedShort()
                require(applicationId == MessageComposer.APPLICATION_AES_ENCRYPTED_PRIVATE_KEY_TRANSFER) { "wrong applicationId: $applicationId" }

                // (2) alias
                val alias = dataInputStream.readString()
                Log.d(TAG, "alias: $alias")
                keyAlias = alias

                // --- AES parameter ---

                // (3) salt
                val salt = dataInputStream.readByteArray()
                Log.d(TAG, "salt: " + byteArrayToHex(salt))

                // (4) IV
                val iv = dataInputStream.readByteArray()
                Log.d(TAG, "IV: " + byteArrayToHex(iv))

                // (5) AES transformation index
                val aesTransformation: AesTransformation =
                    AesTransformation.entries[dataInputStream.readUnsignedShort()]
                Log.d(TAG, "cipher algorithm: $aesTransformation")

                // (6) key algorithm index
                val aesKeyAlg: String =
                    AesKeyAlgorithm.entries[dataInputStream.readUnsignedShort()].keyAlgorithm
                Log.d(TAG, "AES key algorithm: $aesKeyAlg")

                // (7) key length
                val aesKeyLength = dataInputStream.readUnsignedShort()
                Log.d(TAG, "AES key length: $aesKeyLength")

                // (8) AES iterations
                val iterations = dataInputStream.readUnsignedShort()
                Log.d(TAG, "iterations: $iterations")

                // --- Encrypted part ---

                // (9) cipher text
                val cipherText = dataInputStream.readByteArray()
                Log.d(TAG, cipherText.size.toString() + " bytes cipher text read")

                encryptedPayload = EncryptedPayload(
                    salt = salt,
                    iv = iv,
                    cipherText = cipherText,
                    aesTransformation = aesTransformation,
                    keyAlg = aesKeyAlg,
                    keyLength = aesKeyLength,
                    iterations = iterations
                )
            }
        } catch (ex: IOException) {
            printStackTrace(ex)
            Toast.makeText(
                getContext(),
                Objects.requireNonNullElse<String?>(ex.message, ex.javaClass.getName()),
                Toast.LENGTH_LONG
            ).show()
            discardBackStack(this)
        }
    }

    private fun onDecryptClicked() {
        val payload = encryptedPayload ?: return
        Thread {
            onPasswordEntry(payload)
        }.start()
    }

    private fun onPasswordEntry(payload: EncryptedPayload) {
        try {
            //try decrypt
            val aesKeyMaterial = getAesKeyMaterialFromPassword(
                passphrase.toCharArray(),
                payload.salt,
                payload.keyAlg,
                payload.keyLength,
                payload.iterations
            )

            val bArr = process(
                Cipher.DECRYPT_MODE,
                payload.cipherText,
                aesKeyMaterial,
                payload.iv,
                payload.aesTransformation
            )

            Arrays.fill(aesKeyMaterial, 0.toByte())

            try {
                OmsDataInputStream(ByteArrayInputStream(bArr)).use { dataInputStream ->
                    // (9.1) - private key material
                    val privateKeyMaterial = dataInputStream.readByteArray()

                    // (9.2) - public key material
                    val publicKeyMaterial = dataInputStream.readByteArray()

                    val publicKey = restorePublicKey(publicKeyMaterial)
                    val fingerprintBytes = getFingerprint(publicKey)
                    val fingerprintString = byteArrayToHex(fingerprintBytes)
                    requireContext().mainExecutor.execute {
                        decryptedPrivateKeyMaterial = privateKeyMaterial
                        decryptedPublicKeyMaterial = publicKeyMaterial
                        decryptedFingerprintBytes = fingerprintBytes
                        fingerprint = fingerprintString
                        saveEnabled = true
                        validateAlias(fingerprintBytes)
                    }
                }
            } catch (ex: IOException) {
                printStackTrace(ex)
            }
        } catch (ex: Exception) {
            printStackTrace(ex)
            requireContext().mainExecutor.execute {
                Toast.makeText(
                    this.context,
                    "Could not decrypt. Wrong passphrase?",
                    Toast.LENGTH_LONG
                ).show()
            }
        }
    }

    private fun onSaveClicked() {
        if (!saveEnabled) return

        val fingerprintBytes = decryptedFingerprintBytes ?: return
        val privateKeyMaterial = decryptedPrivateKeyMaterial ?: return
        val publicKeyMaterial = decryptedPublicKeyMaterial ?: return

        Thread {
            try {
                //delete other keys with the same fingerprint
                val sameFingerprint = cryptographyManager.getByFingerprint(
                    fingerprintBytes,
                    preferences
                )
                if (sameFingerprint != null) {
                    cryptographyManager.deleteKey(sameFingerprint.alias, preferences)
                }

                val alias = keyAlias.trim { it <= ' ' }
                cryptographyManager.importRsaKey(
                    preferences,
                    privateKeyMaterial,
                    publicKeyMaterial,
                    alias
                )

                //neuer Keystore
                cryptographyManager.importRsaKey(
                    requireActivity().getPreferences(Context.MODE_PRIVATE),
                    privateKeyMaterial,
                    publicKeyMaterial,
                    alias
                )

                requireContext().mainExecutor.execute {
                    Toast.makeText(
                        this.context,
                        "Private key '$alias' successfully imported",
                        Toast.LENGTH_LONG
                    ).show()
                    discardBackStack(this)
                }
            } catch (ex: Exception) {
                printStackTrace(ex)
                requireContext().mainExecutor.execute {
                    Toast.makeText(
                        this.context,
                        ex.message,
                        Toast.LENGTH_LONG
                    ).show()
                }
            }
        }.start()
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        return ComposeView(requireContext()).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                OneMoreSecretTheme {
                    KeyImportScreen(
                        alias = keyAlias,
                        passphrase = passphrase,
                        fingerprint = if (fingerprint.isNotBlank()) {
                            "…%s".format(fingerprint.takeLast(10))
                        } else {
                            getString(R.string.unknown_decrypt_first)
                        },
                        warning = warning,
                        saveEnabled = saveEnabled,
                        onAliasChange = {
                            keyAlias = it
                            validateAlias(decryptedFingerprintBytes)
                        },
                        onPassphraseChange = { passphrase = it },
                        onDecrypt = { onDecryptClicked() },
                        onSave = { onSaveClicked() }
                    )
                }
            }
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        requireActivity().removeMenuProvider(menuProvider)
    }

    private fun validateAlias(fingerprintNew: ByteArray?) {
        try {
            val alias = keyAlias
            Log.d(TAG, "alias: $alias")
            var warningText: String? = null

            cryptographyManager.getByAlias(alias, preferences)?.let {
                if (!it.fingerprint.contentEquals(fingerprintNew)) {
                    warningText = getString(R.string.warning_alias_exists)
                        .format(
                            byteArrayToHex(it.fingerprint).takeLast(
                                10
                            )
                        )
                }
            }

            fingerprintNew?.let {
                val sameFingerprint =
                    cryptographyManager.getByFingerprint(it, preferences)
                if (sameFingerprint != null) {
                    warningText = getString(R.string.warning_same_fingerprint).format(
                        sameFingerprint.alias
                    )
                }
            }

            requireContext().mainExecutor.execute {
                warning = warningText ?: ""
            }
        } catch (e: Exception) {
            printStackTrace(e)
        }
    }

    private inner class KeyImportMenu : MenuProvider {
        override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
            menuInflater.inflate(R.menu.menu_help, menu)
        }

        override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
            if (menuItem.itemId == R.id.menuItemHelp) {
                openUrl(R.string.key_import_md_url, requireContext())
            } else {
                return false
            }
            return true
        }
    }

    companion object {
        private val TAG: String = KeyImportFragment::class.java.getSimpleName()
    }

    private data class EncryptedPayload(
        val salt: ByteArray,
        val iv: ByteArray,
        val cipherText: ByteArray,
        val aesTransformation: AesTransformation,
        val keyAlg: String,
        val keyLength: Int,
        val iterations: Int
    ) {
        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is EncryptedPayload) return false

            if (keyLength != other.keyLength) return false
            if (iterations != other.iterations) return false
            if (!salt.contentEquals(other.salt)) return false
            if (!iv.contentEquals(other.iv)) return false
            if (!cipherText.contentEquals(other.cipherText)) return false
            if (aesTransformation != other.aesTransformation) return false
            if (keyAlg != other.keyAlg) return false

            return true
        }

        override fun hashCode(): Int {
            var result = keyLength
            result = 31 * result + iterations
            result = 31 * result + salt.contentHashCode()
            result = 31 * result + iv.contentHashCode()
            result = 31 * result + cipherText.contentHashCode()
            result = 31 * result + aesTransformation.hashCode()
            result = 31 * result + keyAlg.hashCode()
            return result
        }
    }
}
